hazelcast/src/main/java/com/hazelcast/internal/cluster/impl/MembershipManager.java

```diff
@@ -1268,5 +1385,48 @@ private void innerRun() {
             }
         }
     }
+
+    private class ResolvePartialDisconnectionsTask
+            implements Runnable {
+
+        final Map<MemberImpl, Set<MemberImpl>> disconnections;
+
+        ResolvePartialDisconnectionsTask(Map<MemberImpl, Set<MemberImpl>> disconnections) {
+            this.disconnections = disconnections;
+        }
+
+        @Override
+        public void run() {
+            try {
+                Collection<MemberImpl> membersToRemove = partialDisconnectionHandler.resolve(disconnections);
+                clusterServiceLock.lock();
+                try {
+                    for (MemberImpl member : membersToRemove) {
+                        if (getMember(member.getAddress(), member.getUuid()) == null) {
+                            logger.warning("Won't remove partially disconnected members: " + membersToRemove);
+                            return;
+                        }
+                    }
+
+                    for (MemberImpl member : membersToRemove) {
+                        String reason = format("Removing %s because it has disconnected from some of the members!", member);
+                        logger.warning(reason);
+                        suspectMember(member, reason, true);
+                    }
+                } finally {
+                    clusterServiceLock.unlock();
+                }
+            } catch (Exception e) {
+                logger.severe("Partial disconnection resolution algorithm failed.", e);
```

Comment:
Shouldn't we filter out algorithm timeout? Due to my understanding, this is not always a severe failure...